src/test/scala/com/rawlabs/das/postgresql/PostgresqlBackend.scala

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright 2024 RAW Labs S.A.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0, included in the file
+ * licenses/APL.txt.
+ */
+
+package com.rawlabs.das.postgresql
+
+import java.sql.{Connection, DriverManager, SQLException}
+
+import scala.jdk.CollectionConverters.IteratorHasAsScala
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.rawlabs.das.sdk.DASExecuteResult
+import com.rawlabs.das.sdk.scala.DASTable.TableEstimate
+import com.rawlabs.protocol.das.v1.tables._
+import com.rawlabs.protocol.das.v1.types._
+
+/**
+ * A simple example backend for PostgreSQL, analogous to the SqliteBackend example.
+ *
+ * @param url The JDBC connection string (e.g., "jdbc:postgresql://localhost:5432/mydb")
+ * @param user The database user
+ * @param password The password for that user
+ */
+class PostgresqlBackend(url: String, user: String, password: String, schema: String) {
+
+  /**
+   * Returns a map of tableName -> DASPostgresTable, each containing a TableDefinition. Similar to the original
+   * SqliteBackend#tables().
+   */
+  Class.forName("org.postgresql.Driver")
+  def tables(): Map[String, DASPostgresqlTable] = {
+    val tables = scala.collection.mutable.Map.empty[String, DASPostgresqlTable]
+    var conn: Connection = null
+    try {
+      conn = DriverManager.getConnection(url, user, password)
+
+      // 1) List all user tables in the "public" schema (or adapt if you want all schemas).
+      //    We’ll skip system tables or other schemas by default.
+      val listTablesSQL =
+        s"""
+          |SELECT tablename
+          |FROM pg_tables
+          |WHERE schemaname = '$schema'
+        """.stripMargin
+
+      val stmt = conn.createStatement()
+      try {
+        val rs = stmt.executeQuery(listTablesSQL)
+        try {
+          while (rs.next()) {
+            val tableName = rs.getString("tablename").toLowerCase
+            println(s"Table: $tableName")
+
+            val descriptionBuilder = TableDefinition
+              .newBuilder()
+              .setTableId(TableId.newBuilder().setName(tableName))
+              .setDescription(s"Table $tableName in $schema schema")
+
+            // 2) For each table, list columns using the information schema.
+            //    Retrieve name, type, and nullability.
+            val columnsSQL =
+              s"""
+                 |SELECT column_name, data_type, is_nullable
+                 |FROM information_schema.columns
+                 |WHERE table_schema='public' AND table_name='$tableName'
+               """.stripMargin
+
+            val stmt2 = conn.createStatement()
```

Comment:
Recommend splitting this inner code to a separate method for clarity and avoid stmt2 and rs2 which are harder to follow

Reply:
Done.